CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+behavior and predictable failure modes. Beyond this, we also aim to substantially improve the scalability in order to
+be able to easily support a cluster with thousands of shards.
+
+## Design
+
+### Terminology
+
+| Term | Description |
+| --- | --- |
+| Node | A Single redis-server process |
+| Shard | A logical entity comprised of a group of nodes owning the same slots. It is automatically allocated when a new unique slot range is assigned to a node and automatically deallocated when all slots are moved away |
+| Host | A physical / virtual machine hosting the Flotilla processes |
+
+### High Level Overview
+
+Flotilla is composed of two tiers of strongly consistent, consensus-based control plane systems - The Topology Director 
+(TD) and the Failover Coordinator (FC).
+
+The Topology Director is the source of truth for shard existence and metadata, slot range allocation and Failover Coordinator
+assignments. In addition to the above, it also propagates failover information across the various Failover Coordinators.
+
+The Failover Coordinators is the source of truth for Primary/Replica roles. Each Failover Coordinators is assigned a subset 
+of the total nodes in the cluster, receives & propagates heartbeat information from them and enables them to make failover decisions in a
+consistent & timely manner.
+
+Maintaining these two tiers allows scaling the number of nodes while:
+- Maintaining an upper bound over the time required to decide and propagate failover decisions.
+While the time required to propagate the updated primary/replica status depends on the number of Failover Coordinators in the deployment
+the maximum number of these should always be small enough to allow us to guarantee a substantially better SLA then what
+the current implementation provides.
+- Limiting the clustering overhead on individual nodes in the deployment.
+Depending on pub/sub mode, an individual node only needs to maintain a connection to its Failover Coordinator (and primary/replicas).
+
+While Flotilla can be implemented using any sort of strongly consistent system, the reference implementation we propose
+uses Redis to host the Topology Director & Failover Coordinator logic. This allows for different deployment modes depending on 
+the needs and orchestrating abilities of the administrator:
+- Full Blown: Dedicated nodes holding the Topology Directors and multiple Failover Coordinators dividing the data nodes among them.
+- Midrange: Dedicated nodes holding both the Topology Director and Failover Coordinator (with just one Failover Coordinator owning all the data nodes).
+- Lightweight: Topology Director and Failover Coordinator located on a subset of data nodes. This removes the need to maintain and orchestrate
+non-data nodes at the cost of potentially impacting performance on the dual-use nodes.
+- Naive: Topology Director and Failover Coordinator located on all nodes. This allows for the quickest bootstrapping, but severely limits scalability.
```

Comment:
There's no pre-requisite to select between the two when bootstrapping the original node. As long as all nodes you add are TD+FC+Data you're in Naive mode, then when you add a plain data node you'd be in lightweight.

Admin practices for the two would differ, so the docs make the distinction.

Reply:
for the record, i'm not sure one less configuration would create less confusion. IMHO in some cases if there's something that's technically possible but not supported, that can cause confusion too (and then someone will ask to add it).

Reply:
> for the record, i'm not sure one less configuration would create less confusion. IMHO in some cases if there's something that's technically possible but not supported, that can cause confusion too (and then someone will ask to add it).

But if it's strictly a worse configuration, we'll just say no? 

> There's no pre-requisite to select between the two when bootstrapping the original node. As long as all nodes you add are TD+FC+Data you're in Naive mode, then when you add a plain data node you'd be in lightweight.

I don't agree with the premise that during boot strap you should have to "pick" nodes. A tenet I mentioned earlier was that it should be "As easy as cluster mode" to setup. Why can't it just by default select 3 nodes, and then we disallow going about some number, say 11. Once it exceeds that, it stops adding nodes to the cluster?

Reply:
Might be some confusion here.
There won't be anything inside the TD/FC which says "I'm in [NAIVE/MIDRAGE/...] mode" -- the possibilities are specified to make sure we consider the ramification of them in the rest of the design.

There will be a reference to the various deployment modes in admin tools like redis-cli. This way when redis-cli is told to bootstrap a cluster and provided with 5 nodes it will configure them all as TD+FC+Data by default.

Reply:
Added a clarifying comment